### ISSUE TITLE
from_array / from_delayed etc. with name can cause unexpected side-effects

### ISSUE THREAD

**Issue by MatthiasKohl** [dask/dask#6036]
I'm unsure whether this should be a bug or a feature request, but in any case, I wanted to bring your attention to this behavior which I would call unexpected and hard to debug at the very least.
If it starts a discussion about the design here, that would be great, if not, feel free to close this issue anytime, I know what the problem is and how to work around it.

Here's a snippet of code:
```python
import numpy as np
import dask.array as da
a1 = np.array(list(range(4))).reshape((2, 2))
x1 = da.from_array(a1, chunks=(1, 2), name='input')
out1 = x1.compute()
print('2x2:', out1)
del out1, a1, x1

a2 = np.array(list(range(8))[::-1]).reshape((4, 2))
x2 = da.from_array(a2, chunks=(2, 2), name='input')
out2 = x2.compute()
print('4x2:', out2)
```
Output:

    2x2: [[0 1]
     [2 3]]
    4x2: [[7 6]
     [5 4]
     [3 2]
     [1 0]]

So far so good. Let's try the following:
```python
import numpy as np
import dask.array as da
from dask.distributed import Client
client = Client()
a1 = np.array(list(range(4))).reshape((2, 2))
x1 = da.from_array(a1, chunks=(1, 2), name='input')
out1 = x1.compute()
print('2x2:', out1)
del out1, a1, x1

a2 = np.array(list(range(8))[::-1]).reshape((4, 2))
x2 = da.from_array(a2, chunks=(2, 2), name='input')
out2 = x2.compute()
print('4x2:', out2)
```
Output: well it's complicated, it's not really deterministic (and probably depends a lot on the system you're testing this on). On my Windows 10 laptop with conda, sometimes, I get the above, but sometimes I get this:

    2x2: [[0 1]
     [2 3]]
    4x2: [[0 1]
     [2 3]]

So if you're familiar with the inner workings enough, you might immediately spot the issue:
I'm giving the same *name* to both arrays here which ultimately means that both graphs share some *key*, which is then used by the distributed workers in order to cache objects, and once we go into the second graph, that cache is still there and we get the results from the previous graph.

However, if you're not familiar with said inner workings, I would argue this is a hard issue to spot and honestly the kind of thing that makes me want to never use a library again as it is hard to debug and (in my opinion) very much unexpected behavior.
These are the kinds of bugs I want to avoid the most, as I'll lose a lot of time for something where it simply turns out I didn't understand the interface correctly.

Finally, a few comments and possibilities to address this:

1. This discussion applies in the exact same way to `from_array` as it does to `from_delayed` etc. Whether or not `from_array(..., name=some_name)` is an actual use-case doesn't matter so much since mostly, I want to be able to apply a descriptive name here, rather than gaining time or something like that. So simply using `name=False` or `name=None` is not a solution for me. Plus, all of this applies to `from_delayed` as well where I think descriptive names are important.
2. I would not call this parameter *name* as it clearly doesn't represent a *name*, but a *key*. I would (probably) not have run into this problem if this was called a *key* . So one way to solve this for me would be to simply rename this parameter to *key* or *dask_key_name* as in `dask.Delayed.__call__`.
3. In some sense, I can be happy that I ran into this problem because I got to learn much more about the inner design of the scheduler etc.: whether or not a second dask graph using the same key should expect that key to return the result of a previous graph is up to debate I guess, but I think it doesn't suit the task specification (https://docs.dask.org/en/latest/spec.html) particularly well: in Python, a dict key in one dictionary usually never refers to the value of the same key in another dictionary. So another way to solve this for me would be to clear the worker caches (release_key in distributed worker.py I guess) and if there are similar client/scheduler caches, clear those as well between different graphs.
4. Another way to solve it would be something in between the behavior of `from_array(..., name=False)` and `from_array(..., name=literal_name)`, where either a random GUID or a token taken from the considered object is appended to the name.
5. Finally, I'm surprised this issue never came up before, or at least I think so as I searched for it a while. This applies at least to the 2.9-2.12 versions of Dask/distributed. The only relevant issue similar (but not similar enough) to this one I found is https://github.com/dask/dask/issues/829 from 2015

EDIT1: added `import numpy as np` to the snippets for completeness

**Comment by TomAugspurger** [comment]
Thanks @MatthiasKohl. Do you have a proposal for what to change? I'm not sure that changing the semantics of `name` is an option. It's possible that the only option is better documentation. Currently we have

```
name : str, optional
    The key name to use for the array. Defaults to a hash of ``x``.
    By default, hash uses python's standard sha1. This behaviour can be
    changed by installing cityhash, xxhash or murmurhash. If installed,
    a large-factor speedup can be obtained in the tokenisation step.
    Use ``name=False`` to generate a random name instead of hashing (fast)
```

How would you improve that?

**Comment by MatthiasKohl** [comment]
I think it's not only the documentation that is the issue here. If you take the following:

```python
def inc(x): return x+1
def sum_array(x): return x.sum()
dsk1 = {'x': np.array([1,2]), 'tmp': (inc, 'x'), 'out': (sum_array, 'tmp')}
dsk2 = {'x': np.array([1,2,3]), 'tmp': (inc, 'x'), 'out': (sum_array, 'tmp')}
<scheduler>.get(dsk1, ['out'])
<scheduler>.get(dsk2, ['out'])
```

Then 1) depending on the scheduler, you may not get the same result and 2) the distributed scheduler actually has undefined behavior here as far as I understand it.

So the solution is to 1) ensure that the behavior is defined (as this is what I believe the vast majority of Python users would expect in the example above) and 2) ensure that all schedulers implement that defined behavior.

I can see that if you remove the caching behavior of the distributed scheduler, this could have a huge performance impact. However, is it really obvious to you that the behavior is undefined in the above example? I think you should at least be able to tell the distributed client (or cluster) whether you want to keep the cache between computations. Or you should be able to specify in the dask graph a reference to another graph, which would then either fail or resolve key conflicts based on some heuristic.

**Comment by TomAugspurger** [comment]
I guess my question is how should Dask know that it's doing something wrong? By setting `name`, you're telling Dask to use these values when it encounters that `name` in a task graph.

> I think you should at least be able to tell the distributed client (or cluster) whether you want to keep the cache between computations.

Generally speaking, the Dask scheduler follow's Python's memory management semantics. When all references to the object go away, the result will be freed from the cluster's memory. See https://distributed.dask.org/en/latest/memory.html for more.

However, it is a distributed system. You're original example occasionally shows indeterminate results because sometimes you request `x2` before `x1` has been released, and sometimes after. In this case you can check that `len(client.cluster.scheduler.tasks)` is 0 before proceeding, or more generally ensuring that `out` and all it's dependencies are freed.

> Or you should be able to specify in the dask graph a reference to another graph, which would then either fail or resolve key conflicts based on some heuristic.

This sounds complicated. Typically we merge task graphs prior to computing.

**Comment by MatthiasKohl** [comment]
> However, it is a distributed system

**Comment by jrbourbeau** [comment]
I'm personally in favor of updating the description of `name` to mention that it should be a unique identifier

**Comment by MatthiasKohl** [comment]
> However, it is a distributed system.

**Comment by MatthiasKohl** [comment]
> However, it is a distributed system.

Isn't the whole point of a system like Dask to hide these kind of complexities of a distributed system from the user?
Requesting different computations from Dask where I would like to use the same descriptive names for certain tasks is a perfectly valid use case in my opinion. So why do I have to either ensure that Dask's descriptive name handling (https://github.com/dask/dask/blob/master/dask/optimization.py#L887-L938) works as intended with `tokenize`, or be aware of the cluster's memory model? I think this requires far too intricate knowledge of the library for something that seems fairly common.
It might not be easy to add something like descriptive names to the task model, but I believe it would have a lot of value (for debugging, visualizing, etc.). That's why I wanted to start the discussion.

**Comment by TomAugspurger** [comment]
> Requesting different computations from Dask where I would like to use the same descriptive names for certain tasks is a perfectly valid use case in my opinion

That's certainly reasonable, and it's what Dask does internally. Our tokens typically take the form `{descriptor}-{token}`. Perhaps this pattern would work for you?

```python
In [7]: from dask.base import tokenize

In [8]: a1 = np.array(list(range(4))).reshape((2, 2))
   ...: x1 = da.from_array(a1, chunks=(1, 2), name='input-' + tokenize(a1))
   ...: a2 = np.array(list(range(8))[::-1]).reshape((4, 2))
   ...: x2 = da.from_array(a2, chunks=(2, 2), name='input-' + tokenize(a2))
```

**Comment by TomAugspurger** [comment]
@MatthiasKohl this was auto-closed with #6040, but do take a look at the new docs there and LMK if there's further clarifications.

And if you have other non-documentation suggestions we're open to hearing them, but I think they'll need to work within the constraints of the current Task Graph spec (https://docs.dask.org/en/latest/graphs.html.
